### DISCUSSION TITLE
Hardware guide

### DISCUSSION THREAD

**Discussion by juanitoddd** [CarlosDerSeher/snapclient#27] [General]
Hello friends !
First of all, thank you for keep working on this project 🙏

I would like to kindly ask you for guidance on which hardware to buy, to start this project.
 
I am aware this is based on the ESP32 but what about specific versions/models and also which audio module would be appropriate to make it work. 
I have followed posts about the **MAX98357** and **ma12070p** but not sure which to buy. anyone able to share any experience with them ?

Thank you in advance

**Comment by velaar** [comment]
Mine is running on LyRaT v4.3 boards. 3 boards deployed so far.On Mar 15, 2023, at 9:57 AM, Juan David ***@***.***> wrote:﻿
Hello friends !
First of all, thank you for keep working on this project 🙏
I would like to kindly ask you for guidance on which hardware to buy, to start this project.
I am aware this is based on the ESP32 but what about specific versions/models and also which audio module would be appropriate to make it work.
I have followed posts about the MAX98357 and ma12070p but not sure which to buy. anyone able to share any experience with them ?
Thank you in advance

**Comment by CarlosDerSeher** [comment]
If you have a look at https://github.com/CarlosDerSeher/snapclient/tree/master/components/custom_board you'll find the supported audio codecs there. Also all espressif built devBoards should work.

You can use any esp32 wroom/wrover module based board, but be aware when using wroom based boards you don't have PSRAM available and you will be limited in buffer size (< 750ms) on snapserver.

**Comment by Mannshoch** [comment]
Sorry to ask. Do you think this is doable for someone that first time work with flashing and soldering ESP?
I got a soldering station and a nice course as birthday present and now I wish to try creating room speaker my self.

From this Discoussion it is not clear to me what Hardware I have to buy and the link is not helpfull.
My goal is to listen different Internet radio per room and use it with Kodi (PVR) in the living room. 


Could someone or several please made an example shoping list (from power-suply to speaker and cover) 
Either with detailed product ID or links to the product.

**Reply by CarlosDerSeher** [reply]
First you have to define which output power the audio codec should be able to supply and which impedance the speaker will have. Then have a look at the supported boards and/or DACs and check the datasheets to see if they match your requirements. For small loads you could have a look at MAX98357 and if you need more power maybe TAS5805M. I didn't check all the others.
There are [DIY speaker kits](https://www.hifisound.de/Do-it-yourself-Products/speaker-kits/) you can buy. Just be sure everything will fit inside.

**Comment by anabolyc** [comment]
Hey @juanitoddd 

I'm clearly biased, but I think I'm making the best hardware for this kind of project considering the value for the price :)
I have a range of boards that not only work with snapcast but can be flashed via a [web-installer](https://sonocotta.github.io/esparagus-snapclient/) in minutes since I created pre-built bianries to support them.

On the light side of the range there are these three boards:
|  | [HiFi-ESP32](https://www.tindie.com/products/sonocotta/hifi-esp32/) | [Loud-ESP32](https://www.tindie.com/products/sonocotta/loud-esp32/) | [Louder-ESP32](https://www.tindie.com/products/sonocotta/louder-esp32/) |
|---|---|---|---|
|  | ![DSC_0002](https://github.com/user-attachments/assets/d025ef42-daa3-4ad0-aeec-6ddf309c267b) | ![DSC_0002](https://github.com/user-attachments/assets/7d351d02-7e7f-4974-9a12-ee353c239654) | ![DSC_0013](https://github.com/sonocotta/esp32-audio-dock/assets/5459747/46ea4f10-c363-4623-b5d6-29e1135e5983) |
| MCU | ESP32-WROVER-N16R8 | ESP32-WROVER-N16R8 | ESP32-WROVER-N16R8 |
| DAC | [PCM5100A](https://www.ti.com/product/PCM5100A) 32bit Stereo DAC   -100 dB typical noise level | Dual I2S DAC ([MAX98357](https://www.analog.com/en/products/max98357a.html)) with built in D-Class amp | Stereo I2S DAC ([TAS5805M](https://www.ti.com/product/TAS5805M)) with   built in D-Class amp |
| Output (4Ω) | Non-amplified stereo output, 2.1V RMS | 2x 5W | 2x 32W (4Ω, 1% THD+N) |
| Output (8Ω) | Non-amplified stereo output | 2x 3W | 2x 22W (8Ω, 1% THD+N) |
| PSRAM | 8MB PSRAM (4MB usable) over 40MHz SPI | 8MB PSRAM (4MB usable) over 40MHz SPI | 8MB PSRAM (4MB usable) over 40MHz SPI |
| Power | 5V over USB-C, 2x [LP5907](https://www.ti.com/lit/ds/symlink/lp5907.pdf) 3.3 V Ultra-Low-Noise LDO for analog section | 5V (up to 2.5A) from USB-C | Up to 26V from external PSU   5V over USB-C with power limited to 2x5W |
| Connectivity | WiFi + BT4.2 + BLE   [W5500](https://www.wiznet.io/product-item/w5500/) Ethernet (optional module) | WiFi + BT4.2 + BLE   [W5500](https://www.wiznet.io/product-item/w5500/) Ethernet (optional module) | WiFi + BT4.2 + BLE   [W5500](https://www.wiznet.io/product-item/w5500/) Ethernet (optional module) |

On the heavier side of things, there are 3 similar products with a bit more capabilities and a bit more peripheral

|  | [HiFi Esparagus](https://www.tindie.com/products/sonocotta/esparagus-hifi-medialink/) | [Loud Esparagus](https://www.tindie.com/products/sonocotta/loud-esparagus-media-center/) | [Louder Esparagus](https://www.tindie.com/products/sonocotta/louder-esparagus-media-center/) |
|---|---|---|---|
| Image | ![DSC_0709](https://github.com/sonocotta/esparagus-hifi-medialink/assets/98712315/ea45f1d2-32b5-4f12-a63c-a8e403cb22db) | ![DSC_0706](https://github.com/sonocotta/esparagus-hifi-medialink/assets/5459747/2556b8ff-1827-4e03-8e28-31e40199943c) | ![DSC_0713](https://github.com/sonocotta/esparagus-media-center/assets/5459747/14d54647-2b7e-4b1a-9a8e-135a1598eb02) |
| MCU | ESP32-WROVER-N16R8 | ESP32-WROVER-N16R8 | ESP32-WROVER-N16R8 |
| DAC | PCM5100A 32bit Stereo DAC -100 dB typical noise level | Dual I2S DAC ([MAX98357](https://www.analog.com/en/products/max98357a.html)) with built in D-Class amp | Stereo I2S DAC ( [TAS5805M](https://www.ti.com/product/TAS5805M) ) with built in D-Class amp |
| Power | 5V over USB-C, 2x [LP5907](https://www.ti.com/lit/ds/symlink/lp5907.pdf) 3.3 V Ultra-Low-Noise LDO for analog section | 5V from USB-C | Up to 20V from USB-C PD or Up to 26V from external PSU |
| Output, 4Ω | Non-amplified stereo output | 2x 3W | 2x 32W (4Ω, 1% THD+N) |
| Output, 8Ω | - | 2x 5W | 2x 22W (8Ω, 1% THD+N) |
| PSRAM | 8MB PSRAM (4MB usable) | 8MB PSRAM (4MB usable) | 8MB PSRAM (4MB usable) |
| Peripheral | WS2812B RGB Led   SSD1306 128x64 OLED screen (optional) | WS2812B RGB Led   SSD1306 128x64 OLED screen (optional) | WS2812B RGB Led   SSD1306 128x64 OLED screen (optional) |
| Size | 80 x 50 x 20mm | 80 x 50 x 20mm | 100 x 80 x 38mm |

My persoanl favourite is a TAS5805M based devices, since this DAC has extremely powerful DSP that I'm working on right now to provide a good driver for. I added a few test configurations in my [snapcast fork](https://github.com/sonocotta/esparagus-snapclient/), You can find them on the installer page linked above, but this is just the tip of the iceberg, and I'm working to provide more granular control on the DSP possibilities.

Please don't hesitate to ask any question you might have.

**Reply by Wookbert** [reply]
Wow, just wow! These are exactly the types of boards I’ve been looking for: an ESP32 for use as SnapCast client next to a real amplifier, powerful enough to drive kitchen or bath room speaker (and a little beyond). Awesome find...

@anabolyc  I intend to go for the Louder-ESP32 and noticed that you offer on Tindie versions with ESP32-WROVER-N16R8 or ESP32-S3-WROOM-N16R8. Which one to pick? Where are the differences?

This ESP32 SnapCast Client installs and runs flawlessly (I assume installation via the above mentioned [web-installer](https://sonocotta.github.io/esparagus-snapclient/))? How’s power consumption when muted, so just the ESP32 receiving the stream, but not outputting it/the amplifier doing nothing? How fast does the amp react to a volume change which is triggered externally, e.g. via Home Assistant? (I’d like to built a separate, Wifi-attached controller with round ESP32-OLED-Display-Module for title info and an attached rotary knob to adjust volume on your Louder-ESP32 board).

**Reply by anabolyc** [reply]
Hey @Wookbert 

S3 would only work with the esphome port at the moment, standalone snaplient only works with classic esp32. Regarding power consumption, i need to double check, but i think snapclient uses afd driver, so it would go mute mode when idling, so we’re talking just few milliamps. More details here: https://github.com/sonocotta/esp32-tas5805m-dac?tab=readme-ov-file#power-measurements-on-louder-esp32-board
Reaction time depends mostly on the buffer settings, it is always a compromise between delay and playback stability. In my estup (with default settings) it was aroung half second i think.

**Reply by juanitoddd** [reply]
@anabolyc thank you very much sir for this list !. 

Currently, in my setup i have HomeAssitant with MusicAssistant and 4 [squeezelite](https://github.com/sle118/squeezelite-esp32) players. Multiroom audio works but after the 3rd, 4th song, it gets out of sync. Thats why I am looking again to this discussion, I am interested in the `louder-esp32-board` to replace my squeezelite's.

My concrete question, is, would a snapcast setup be more robust when it comes to multi-room audio-sync over squeeze/LMS ? 
Based on this I may replace my setup to snapcast.

**Reply by CarlosDerSeher** [reply]
My first steps towards multiroom audio were with raspberry pi2/3 and LMS too but never could make it work reliable. That's when I found snapcast and was overwhelmed how well it worked. Lots of coding later we came to a mostly stable esp snapclient. So yes, from my opinion you should definitely get down this route.

But I see you use esp32 already, which hardware do you have and is it support by snapclient?

**Reply by juanitoddd** [reply]
@CarlosDerSeher Thank you. Yes, I use 4 units of [SqueezeAMP](https://github.com/philippe44/SqueezeAMP) which have a WROVER board. But I am not sure how compatible would be to flash a snapclient onto them. I believe (as the name suggest) they are exclusively made for squeeze clients. I mean I could try to flash one and see what happens :) . 

Other than that, I am looking to get some [louder-esp32](https://www.tindie.com/products/sonocotta/louder-esp32/) which explicitly support snapclient

**Reply by anabolyc** [reply]
SqueeezeAmp uses MA12070 DAC, I was about to test it with [Loud-ESP32-Plus](https://sonocotta.com/loud-esp32/) (not published yet). If it works, I can add SqueezeAmp config as well, should be only a few pins difference.

**Reply by CarlosDerSeher** [reply]
There is a driver available called ma120x0 could this be used for this Board?

Edit: the header states `// Register interface thrugh I2C for MA12070P and MA12040P`

**Reply by CarlosDerSeher** [reply]
@anabolyc are you sure about DAC the linked repo states `TAS575x stereo class-D amplifier`

**Reply by anabolyc** [reply]
AH. Mixed it up with [muVox](https://github.com/muvox-io/muVox). You're right.
If you give me working config, I can integrate it into web-installer, but no way I can test it.

**Reply by CarlosDerSeher** [reply]
@juanitoddd are you able to build and test yourself? You'll need the gpio pins to which the DAC is connected. Shouldn't be much of a problem to gather this information though
